crates/bevy_color/src/color_gradient.rs

```diff
@@ -0,0 +1,184 @@
+        let colors = colors.into_iter().collect::<Vec<_>>();
+        let len = colors.len();
+        (!colors.is_empty())
+            .then(|| Self { colors })
+            .ok_or(ColorGradientError(len))
+    }
+
+    /// Converts the [`ColorGradient`] to a [`ColorCurve`] which implements the [`Curve`] trait
+    /// along with its adaptor methods
+    /// # Example
+    ///
+    /// ```
+    /// # use bevy_color::palettes::basic::*;
+    /// # use bevy_color::ColorGradient;
+    /// # use bevy_color::Srgba;
+    /// # use bevy_color::Mix;
+    /// # use bevy_math::curve::Curve;
+    /// let gradient = ColorGradient::new([RED, GREEN, BLUE]).unwrap();
+    /// let curve = gradient.to_curve();
+    ///
+    /// // you can then apply useful methods ontop of the gradient
+    /// let brighter_curve = curve.map(|c| c.mix(&WHITE, 0.25));
+    ///
+    /// assert_eq!(brighter_curve.sample_unchecked(0.0), Srgba::new(1.0, 0.25, 0.25, 1.0));
+    /// ```
+    pub fn to_curve(self) -> ColorCurve<T> {
+        let domain =
+            Interval::new(0.0, (self.colors.len() - 1) as f32).expect("at least 1 by construction");
+        ColorCurve {
+            domain,
+            gradient: self,
+        }
+    }
+}
+
+impl<T: Mix> ColorRange<T> for ColorGradient<T> {
+    fn at(&self, factor: f32) -> T {
+        match self.colors.len() {
+            0 => {
+                unreachable!("at least 1 by construction")
+            }
+            1 => {
+                // This weirdness exists to prevent adding a `Clone` bound on the type `T` and instead
+                // work with what we already have here
+                self.colors[0].mix(&self.colors[0], 0.0)
+            }
+            len => {
+                // clamp to range of valid indices
+                let factor = factor.clamp(0.0, (len - 1) as f32);
+                let fract = factor.fract();
+                if fract == 0.0 {
+                    // doesn't need clamping since it already was clamped to valid indices
+                    let exact_n = factor as usize;
+                    // weirdness again
+                    self.colors[exact_n].mix(&self.colors[exact_n], 0.0)
+                } else {
+                    // SAFETY: we know that `len != 0` and `len != 1` here so `len >= 2`
+                    let below = (factor.floor() as usize).min(len - 2);
+                    self.colors[below].mix(&self.colors[below + 1], fract)
+                }
+            }
+        }
+    }
+}
+
+/// Error related to violations of invariants of [`ColorGradient`]
+#[derive(Debug, thiserror::Error)]
+#[error(
+    "Couldn't construct a ColorGradient since there were too few colors. Got {0}, expected >=1"
+)]
+pub struct ColorGradientError(usize);
+
+/// A curve whose samples are defined by a [`ColorGradient`]. Curves of this type are produced by
+/// calling [`ColorGradient::to_curve`].
+#[derive(Clone, Debug)]
+pub struct ColorCurve<T: Mix> {
```

Comment:
This is basically just `SampleCurve` with `<T as Mix>::mix` as far as I understand it.